TicTacToeBot/Source/XOPlayer.cpp

```diff
@@ -0,0 +1,41 @@
+//
+// Created by Sam Petrov on 25.05.2020.
+//
+
+#include "XOPlayer.h"
+#include <cassert>
+
+XOPlayer::XOPlayer(TreeNode &tree, PlayField::csState symbol): Tree(tree) {
+    CurNode = &Tree;
+    selectPlayer(symbol);
+}
+
+void XOPlayer::selectPlayer(PlayField::csState symbol) {
+    Bot = (symbol == PlayField::csCross) ? PlayField::csNought : PlayField::csCross;
+}
+
+void XOPlayer::makeMove(PlayField::CellPos iCell) {
+    assert(CurNode->value()[iCell] == PlayField::csEmpty);
+    for (int i = 0; i < CurNode->childCount(); ++i)
+        if ((*CurNode)[i].value()[iCell] != PlayField::csEmpty) {
+            CurNode = &(*CurNode)[i];
+            break;
+        }
+}
+
+bool XOPlayer::isBestMove(TreeNode &goodTurn, TreeNode &current) const {
+    return (Bot == PlayField::csCross && current.getCrossesWin() > goodTurn.getCrossesWin()
+            || Bot == PlayField::csNought && current.getNoughtsWin() > goodTurn.getNoughtsWin())
+            && current.getDraw() > goodTurn.getDraw();
```

Comment:
Критерий не соответствует ТЗ. :_(

Reply:
`Критерий жадного алгоритма: случайно выбирается один из вариантов, для которого сумма количества выигрышей за бота и количества ничьих наибольшее.`

Reply:
Правильно. Но даже в этой трактовке у вас неправильно написано. А надо было бы так:
```suggestion
    return (Bot == PlayField::csCross && current.getCrossesWin()+current.getDraw() > goodTurn.getCrossesWin()+goodTurn.getDraw()
            || Bot == PlayField::csNought && current.getNoughtsWin()+current.getDraw() > goodTurn.getNoughtsWin()+goodTurn.getDraw());
```
А в конференции мы критерий уточняли, и он выглядит не сейчас так:

> Критерий жадного алгоритма: случайно выбирается один из вариантов, для которого !!отношение!! суммы количества выигрышей за бота и ничьих к общему количеству исходов наибольшее.

Я думаю причины вам понятны.